DFG ctrl flow optimized

HardyVon opened the pull request:
Optimize the generation of DFG.
Changed `isLiveInInst` function in `DFG.cpp`.
Now only consider br related ctrl flow and phi related ctrl flow.

tancheng commented on src/DFG.cpp:
```diff
@@ -1170,6 +1170,23 @@ void DFG::initPipelinedOpt(list<string>* t_pipelinedOpt) {
+
+  // CHANGED.
+  /*
   if (t_inst == &(t_bb->front())) {
```
Well, after reading following code and recalling my memory, I believe the live-in analysis and the ctrl connections are reasonable and should be there... Removing them would break the correctness. Let's sync again and talk about the other optimization opportunities.

HardyVon replied:
Dear Tan，
    After changing `isLiveInInst` function, I can't pass DVFS test. But the problem seems does not lie on `isLiveInInst`. In fact , it lies on your test script `verify.sh`. Just add a `\` in front of `[` and `]`, because `[` and `]` have meaning in `grep` command you used.
    Best,
    Hardy

tancheng replied:
Thanks, you can fix it if you want. However, my comment about the correctness is orthogonal with the testing. I have concern about the missing/removed ctrl edges.

Specifically, the live-in DFGNodes besides `phi` in a basic block should also be controlled by a `br`, unless it's dominator basic block (parent basic block) is already controlled by the same `br`. We can discuss with a concrete example here or during our meeting.

HardyVon replied:
Dear Tan,
    I read carefully about the original `isLiveInInst` function. The logic as follows:
    (a) Any `t_bb.front()` inst is regarded as live-in.
    I recognize this, for it is always `br` related.
    (b) If (a) is not satisfied, any inst uses data generated by another inst that is NOT in the same basicblock as it is regarded as live-in.
    This causes some problems of redundant ctrl flows pointed to `store`, `add`, `mul`, et cetera.
    (c) If (a) and (b) are not satisfied, any inst uses data generated by another inst that is in the same basicblock as it with a lower ID is **NOT** regarded as live-in. 
    I think this is right. Inst of type (c) shouldn't be regarded as live-in.
    (d) If (a), (b) and (c) are all not satisfied, the remaining inst should be regarded as live-in. 
    This also causes some problems of redundant ctrl flows.
    When I run the `nested_loop` exmaple, I noticed that a ctrl flow from `(18)br` to `(3)br` is established if I use the original `isLiveInInst` function, but this ctrl flow disappeared when I use the modified `isLiveInInst` function.
    I noticed that `(3)br`(full inst: `br label %for.body3`) is an inst of type (d). Although `(3)br` is an unconditional jump inst, it doesn't mean that it will trigger in every II. So, `(3)br` should be controlled by `(18)br`, for `(18)br` should "tell" `(3)br` to jump or not. This ctrl flow should not be removed.
    In summary, if there's a "lonely" inst which has no ctrl flow or data flow pointed to it, we should consider to establish a ctrl flow.
    This is the problem I discovered. Is my understanding correct? If there's any problems you find, please let me know.
    Best wishes,
    Hardy

tancheng replied:
> In summary, if there's a "lonely" inst which has no ctrl flow or data flow pointed to it, we should consider to establish a ctrl flow.

Exactly.

> (b) If (a) is not satisfied, any inst uses data generated by another inst that is NOT in the same basicblock as it is regarded as live-in.
This causes some problems of redundant ctrl flows pointed to store, add, mul, et cetera.

 - I have concern for this one. I think the connection for this (b) could guarantee the correctness in a conservative way. And (b) is truely "live-in" based on its meaning from the basic block perspective.
 - Then let's see when it is safe to remove the ctrl connection for (b):
   - I think if any of (b)'s operands is also controlled by the same ctrl connection (i.e., br), we are safe to remove the one connecting to (b). I am not sure whether this is always the case though.
 
WDYT?

HardyVon replied:
Dear Tan,
&nbsp;&nbsp;I think some of (b) and some of (d) should be regarded as live-in, and the rest of (b) and the rest of (d) can be regarded as not live-in, but at the same time we should add a "mark" to the data flows pointed to them.
&nbsp;&nbsp;I'll use sets to illustrate my point. $A$ denotes to the set of insts that satisfy (a), $B, C, D, E, F, G, H$ follows the same principle.
&nbsp;&nbsp;Here (e) (f) (g) (h) are:
&nbsp;&nbsp;&nbsp;&nbsp;(e) $\Leftrightarrow$ (a). So $E = A$.
&nbsp;&nbsp;&nbsp;&nbsp;(f) `phi` inst that not satisfy (e). Obviously $F \subset B$.
&nbsp;&nbsp;&nbsp;&nbsp;(g) "lonely" inst that not satisfy (e) and (f). Obviously $G \subset D$.
&nbsp;&nbsp;&nbsp;&nbsp;(h) the remaining inst that not satisfy (e) (f) (g). So $H = \complement_B F \cup \complement_D G \cup C$.
&nbsp;&nbsp;We used to generate ctrl flow for inst $i_0 \in A \cup B \cup D$. Now I'll only generate ctrl flow for inst $i_1 \in E \cup F \cup G$. 
&nbsp;&nbsp;I'll also add a "mark" to the data flows that makes inst $i_2 \in  \complement_B F \cup \complement_D G$. This mark is designed to "tell" CGRA-Mapper to preserve the register **FOREVER** which stores the data when mapping (i.e. the duration limit of this register is $+\infty$ when mapping). This is because we're not sure when these inter-basicblock data($\complement_B F$ situation) and inter-iteration data($\complement_D G$ situation) will be used, so the registers shouldn't be flushed. By the way, I think the existing CGRA-Mapper doesn't guarantee this.
&nbsp;&nbsp;We don't need to add this mark to $i_3 \in  C$ because $i_3$ only uses intra-iteration data. Also, we don't need to add this mark to $i_1 \in E \cup F \cup G$ because we won't use the `br` result at the later iteration.
&nbsp;&nbsp;This is my preliminary idea. What do you think?
Yours,
Hardy

tancheng replied:
- I didn't get "lonely". You'd better draw a pseudo DFG diagram with pseudo code to illustrate different categories of DFG nodes. I still think "if any of (b)'s operands is also controlled by the same ctrl connection (i.e., br), we are safe to remove the one connecting to (b)".
 - I agree on "mark", it could be a set, i.e., `ctrl_sources = {}`.
 - I didn't get **FOREVER**. First of all, as we mentioned before, the DFG optimization and mapping could be orthogonal. Now, you are talking about how to map DFG based on the "marked" DFG. Then, the value stored in the register should also be conditional updated. **FOREVER** sounds like read-only for me, which doesn't make sense. Actually, I don't think we need this "FOREVER" data, the predicate is already correctly delivered to the target DFG node.

Plz consider "draw a pseudo DFG diagram with pseudo code" so we can discuss during our meeting.

HardyVon replied:
- “lonely” means an inst has neither ctrl flow nor data flow pointed to it. For example, unconditional branch.
- **FOREVER** doesn't mean read-only. The data stored in this register can be and should be conditional updated. I mean, this register is preserved **ONLY** for these inter-basicblock and inter-iteration data. Other data can not occupy this register at any time.
- predicate can achieve correct conditional update. This does not contradict forever register. Predicate is to correctly generate the newest data, while forever register is to make sure that if data update doesn't happen, the FU can still get the right data stored before.
- By the way, the "mark" sign I added can also be a "stop" sign of predicate propagation. Predicate signal should not propagate when it meets a “marked” data flow or a ctrl flow（not include the first ctrl flow）.

HardyVon commented:
- I made some changes in `DFG.cpp`, especially in function `construct` and `isLiveInInst`.
- I also changed the workflow file `cmake.yml`. I added `wokflow_dispatch:` to enable manual running workflow test. But this neeeds your approval, otherwise this PR will not run workflow test.

tancheng commented on src/DFG.cpp:
```diff
@@ -8,1798 +8,1971 @@
+ DFG::DFG(Function& t_F, list<Loop*>* t_loops, bool t_targetFunction,
+          bool t_precisionAware, bool t_heterogeneity,
+          map<string, int>* t_execLatency, list<string>* t_pipelinedOpt,
+    bool t_supportDVFS, bool t_DVFSAwareMapping,
```
The arguments should align with the ones above it.

tancheng commented on src/DFG.cpp:
```diff
@@ -8,1798 +8,1971 @@
+           dfg_node->setDVFSLatencyMultiple(2);
+     assigned_dvfs_nodes.insert(dfg_node);
+     mid_dvfs_dfg_nodes += 1;
+   }
```
The format (indent) seems messed up. Can you make a cleanup PR for the fix? Let's use "two space" rather than "tab".

tancheng commented on src/DFG.cpp:
```diff
@@ -8,1798 +8,1971 @@
+   int num_tiles_in_island = t_DVFSIslandDim * t_DVFSIslandDim;
+   int unused_high_dvfs_cgra_tiles_across_II =
+         t_II * num_tiles_in_island *
+   ((high_dvfs_dfg_nodes + num_tiles_in_island - 1) / num_tiles_in_island) -
```
All these indents are messed up, we should keep what we have before.

tancheng commented on src/DFG.cpp:
```diff
@@ -8,1798 +8,1971 @@
+         unused_mid_dvfs_cgra_tiles_across_II -= 2;
+         unused_low_dvfs_cgra_tiles_across_II -= 1;
+       } else {
+   // If all the islands assuming the optimal II are used up, label
```
Messed up here. I believe it is the IDE's issue, the "tab" and "space" are not interpreted as expected.

tancheng commented on src/DFG.cpp:
```diff
@@ -8,1798 +8,1971 @@
+         DFGNode* dfgNode;
+ //        if (hasNode(curII)) {
+ //          dfgNode = getNode(curII);
+ //        } else {
```
Remove these commented out lines.

tancheng commented on src/DFG.cpp:
```diff
@@ -8,1798 +8,1971 @@
+           dfgNode = new DFGNode(nodeID++, m_precisionAware, curII, getValueName(curII), m_supportDVFS);
+           dfgNode->setBBID(bbID);
+           nodes.push_back(dfgNode);
+ //        }
```
Remove.

tancheng commented on src/DFG.cpp:
```diff
@@ -8,1798 +8,1971 @@
+     }
+   }
+   
+   // construct data flows.
```
Following commented out lines can be removed.

tancheng commented on src/DFGEdge.h:
```diff
@@ -38,6 +41,12 @@ class DFGEdge
     DFGNode* getConnectedNode(DFGNode*);
     bool isCtrlEdge();
+    
+    // set m_isInterEdge.
```
Provide a brief explanation here for `setInterEdge()`.

tancheng commented on src/DFGEdge.h:
```diff
@@ -38,6 +41,12 @@ class DFGEdge
+    // set m_isInterEdge.
+    void setInterEdge(bool);
+
+    // read m_isInterEdge.
```
ditto

tancheng commented on src/DFGNode.cpp:
```diff
@@ -590,7 +590,7 @@ list<DFGNode*>* DFGNode::getPredNodes() {
     m_predNodes->push_back(edge->getSrc());
   }
-  if (isBranch()) {
+  if (isBranch()) { 
```
Is there an additional "space" after `{`?

tancheng commented on src/DFGNode.h:
```diff
@@ -68,6 +68,9 @@ class DFGNode {
     int m_DVFSLatencyMultiple;
     bool m_supportDVFS;
 
+    // "m_bbID" is used to specify which basicblock does this DFGNode in.
```
`does` -> `is`.

tancheng commented on src/DFGNode.h:
```diff
@@ -145,6 +148,12 @@ class DFGNode {
     void setDVFSLatencyMultiple(int);
     int getDVFSLatencyMultiple();
 
+    // set m_bbID.
```
`// Sets m_bbID.`

tancheng commented on src/DFGNode.h:
```diff
@@ -145,6 +148,12 @@ class DFGNode {
+    // set m_bbID.
+    void setBBID(int);
+
+    //read m_bbID.
```
// Reads m_bbID.

tancheng reviewed (commented):
Hi @HardyVon, I left some comments, most about the format/indent. Please try to fix them in another PR for the cleanup. Thanks!
